feat: ws api wrapper

jackstar12 opened the pull request:
Introduces a `BoltzApiWs` struct which simplifies the usage of the websocket.

This is part of an effort for creating https://mozilla.github.io/uniffi-rs/latest/ bindings for the library.
Helpers like this make for easier bindings since one has to expose less functionality to the foreign language

todo:

- [x] tests for `BoltzApiWs`
- [x] update other tests to use `BoltzApiWs`

jackstar12 committed:
feat: ws api wrapper

michael1011 commented on src/swaps/boltz.rs:
```diff
@@ -323,6 +323,11 @@ impl BoltzApiClientV2 {
     }
 
+    pub fn ws(&self) -> BoltzWsApi {
+        let ws_string = self.base_url.clone().replace("http", "ws") + "/ws";
```
Should also handle https

jackstar12 replied:
If you replace `http` in `https` with `ws` you end up with `wss`

michael1011 replied:
What I mean is:
- http:// -> ws://
- https:// -> wss://

jackstar12 replied:
Which this code correcly handles:
`https` will turn into `wss` since the last s is simply left as-is

michael1011 replied:
`replace` replaces *all occurrences*: https://doc.rust-lang.org/std/string/struct.String.html#method.replace
So if someone had `http` in their URL, this would break it

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,183 @@
+
+pub struct BoltzWsApi {
+    ws_url: String,
+    swap_ids: Mutex<Vec<String>>,
```
We could use a HashSet here to avoid duplicate entries

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,183 @@
+                            tokio::select! {
+                                _ = &mut shutdown_receiver => {
+                                    info!("Received shutdown signal, exiting socket loop");
+                                    break;
```
If I am not mistaken this will only break the inner loop, and cause a reconnection. You can use labels to break the outer loop from this line

jackstar12 committed:
fix: break out of connection loop using label

jackstar12 committed:
refactor: use hashset for `swap_ids`

jackstar12 committed:
refactor: update all tests to use new `BoltzApiWs`

jackstar12 committed:
fix: tokio wasm support

jackstar12 committed:
Merge branch 'trunk' into feat/ws-api-wrapper

jackstar12 committed:
fix: use wasm_bindgen_futures to spawn task in wasm

jackstar12 committed:
feat: ws improve implementation with config and restart handlers

jackstar12 committed:
refactor: dont limit regtest tests to regtest folder

tests that require regtest shouldnt be forced into the regtest folder,
to allow for access to private states

jackstar12 commented:
Still having issues with WASM

jackstar12 committed:
fix: make sure to create ws update receiver before subscribing

we might miss the initial swap.created message otherwise

jackstar12 committed:
fix: use gloo_timers for timeouts on wasm

jackstar12 committed:
refactor: leave spawning a task for the ws loop to implementations

theres no straight-forward way to spawn a task on wasm, so we wanna
leave that to the implmentation

michael1011 commented on Cargo.toml:
```diff
@@ -38,36 +38,35 @@ hex = "0.4"
 lnurl-rs = { version = "0.9.0", default-features = false, features = ["async", "async-https"], optional = true }
 # Pin to commit until https://github.com/seanmonstar/reqwest/pull/1760 gets released
 reqwest = { git = "https://github.com/seanmonstar/reqwest", rev = "00b15b9a893d350388af513179e1a973dfa26f85", features = ["json"] }
+tokio = { version = "1.43.0", features = ["macros", "time", "sync", "rt"] }
```
I don't know how i feel about adding tokio by default. It's pretty heavy and not everyone is using it

Might make sense to hide it behind a feature. What do you think @i5hi?

jackstar12 replied:
I will put it behind a `ws` feature

jackstar12 replied:
We already depend on tokio in non-wasm by default in the electrum-client, altough it only uses `time`. In order for a ws feature to make sense, we should also put tokio-tungstenite-wasm behind the same flag, since that also depends on parts of tokio, which would be a breaking change, unless we enable the ws feature by default. 

I suggest:
- depend on tokio time by default
- depend on tokio-tungstenite-wasm and tokio "sync" additionally with the ws feature
- depend on tokio rt and macros in dev

michael1011 commented on src/swaps/boltz.rs:
```diff
@@ -748,9 +753,13 @@ pub enum WsRequest {
 
 impl WsRequest {
     pub fn subscribe_swap_request(swap_id: &str) -> Self {
+        Self::subscribe_swaps_request(vec![swap_id.to_string()])
```
```suggestion
    pub fn subscribe_swap_request(swap_id: String) -> Self {
        Self::subscribe_swaps_request(vec![swap_id])
```
Would be nicer but is breaking

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    async fn wait_for_subscription(
+        &self,
```
This does not need self

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,366 @@
+        swap_id: &str,
+    ) -> Result<(), Error> {
+        // First, wait for the result from the subscribe call
+        response_receiver
+            .await
+            .map_err(|_| Error::Generic("Failed to receive subscription response".to_string()))??;
+
+        // Then, wait for the successful subscription response from boltz
+        loop {
+            if subscriptions.recv().await.map_err(|_| {
+                Error::Generic("Failed to receive subscription notification".to_string())
+            })? == swap_id
+            {
+                return Ok(());
+            }
+        }
```
Isn't that the same thing?

jackstar12 replied:
No. The first error happens if there was an error sending the subscribe message to boltz and the second waits for the subscribe response from boltz

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    async fn sleep(&self, duration: Duration) {
```
This does not need self

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,366 @@
+        match select(wait, timeout_future).await {
+            futures_util::future::Either::Left((result, _)) => result,
+            futures_util::future::Either::Right((_, _)) => Err(Error::Generic(
+                "Subscription timeout, attempting to reconnect".to_string(),
```
This does not always reconnect?

jackstar12 replied:
Wording is confusing here, should only be `Subscription timeout`. The reconnecting is handled by the `subscribe` call (this is `try_subscribe`)

michael1011 commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,366 @@
+
+                            _ = interval.next() => {
+                                match connection.send_json(&WsRequest::Ping).await {
+                                    Ok(_) => debug!("Sent keep-alive ping"),
```
This will print *a lot*. Let's do trace

jackstar12 committed:
chore: cleanup

jackstar12 committed:
feat: introduce sleep util

jackstar12 committed:
Merge branch 'trunk' into feat/ws-api-wrapper

jackstar12 committed:
feat: ws feature

jackstar12 committed:
refactor: use tokio select instead of futures_util select

jackstar12 committed:
chore: improve error handling

danielgranhao commented on src/swaps/status_stream.rs:
```diff
@@ -0,0 +1,346 @@
+        }
+    }
+
+    #[cfg(not(all(target_family = "wasm", target_os = "unknown")))]
+    pub fn start(self: Arc<Self>) {
+        tokio::spawn(self.run_ws_loop());
+    }
```
Why not provide a Wasm alternative based on `spawn_local?`

jackstar12 replied:
We figured that we wanna impose as little as possible restrictions on the implementations of the WASM side, leaving that to the implementation.

Maybe this one should be removed aswell to keep consistency across platforms.

danielgranhao reviewed (approved):
Looks very good. Left just a suggestion

jackstar12 committed:
chore: remove `start` method

we want to leave spawning the task to the implementation, especially on
wasm

jackstar12 committed:
chore: remove unnecessary sleep before reconnect
